### ISSUE TITLE
doom-first-buffer-hook is not triggered by evil-buffer-new

### ISSUE THREAD

**Issue by Janfel** [hlissner/doom-emacs#4514]
**What did you expect to happen?**
The transient hook `doom-first-buffer-hook` is called when I interactively open a new buffer using `evil-buffer-new` bound to `SPC b N`. This results in `smartparens-global-mode` being called.

**What actually happened?**
When I open a new buffer using `SPC b N` in a fresh Emacs instance, `doom-first-buffer-hook` is not run and `smartparens-global-mode` is not started.

**Additional details:**
- [My Emacs Config](https://github.com/Janfel/doom-emacs-config)
- `SPC b N` : `evil-buffer-new`
- `SPC ;` : `pp-eval-expression`

**Steps to reproduce:**
1. Start a fresh Emacs instance.
2. Open a new buffer using `evil-buffer-new` interactively.
3. Evaluate `doom-first-buffer-hook` using `pp-eval-expression` on `SPC ;`. It is not `nil`, indicating it has not been called yet.
4. Evaluate `smartparens-mode` and `smartparens-global-mode` using `pp-eval-expression` on `SPC ;`. Both variables are `nil`.

**System information:**
<details><pre>
SYSTEM  type       gnu/linux
	config     x86_64-pc-linux-gnu
	shell      /usr/bin/zsh
	uname      Linux 5.10.2-2-MANJARO #1 SMP PREEMPT Tue Dec 22 08:14:42 UTC 2020 x86_64
	path       (~/bin ~/opt/bin /usr/local/sbin /usr/local/bin /usr/bin /usr/lib/jvm/default/bin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl ~/.local/bin ~/.dotnet/tools ~/var/go/bin ~/.local/share/cabal/bin ~/opt/ghcup/bin ~/.gem/ruby/2.7.0 ~/.local/share/cargo/bin /usr/lib/emacs/27.1/x86_64-pc-linux-gnu)
EMACS   dir        ~/opt/doom-emacs/
	version    27.1
	build      Aug 28, 2020
	buildopts  --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --localstatedir=/var --with-x-toolkit=gtk3 --with-xft --with-wide-int --with-modules --with-cairo --with-harfbuzz 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fno-plt' CPPFLAGS=-D_FORTIFY_SOURCE=2 LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
	features   XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP
	traits     (batch server-running envvar-file)
DOOM    dir        ~/.config/doom-emacs/
	version    2.0.9
	build      HEAD -> develop, origin/develop, origin/HEAD 0c9256411 2020-12-15 04:14:40 -0500
	elc-files  0
	modules    (:completion (company +childframe) ivy :ui doom doom-dashboard doom-quit hl-todo modeline nav-flash neotree ophints (popup +defaults) vc-gutter window-select workspaces :editor (evil +everywhere) file-templates fold (&user fmt +define) rotate-text snippets :emacs dired electric ibuffer undo vc :term vterm :checkers syntax :tools editorconfig (eval +overlay) lookup lsp magit make :lang data emacs-lisp markdown org sh yaml (&user arduino) (&user basic) (cc +lsp) clojure common-lisp elixir (&user fennel) (haskell +dante) latex (lua +moonscript) (python +lsp) (rust +lsp) scheme web :config (&user literate) (default +evil +bindings +smartparens))
	packages   ((meson-mode) (moonscript :pin ac5d35d3a1970db958c687991a1f2c49f7921150 :recipe (:fork (:host github :repo Janfel/moonscript-mode))) (php-mode) (evil-collection :pin 034124fb09e7a5f57e23bafb08d3d9135d272d90) (async) (flycheck-clang-tidy) (evil-escape :disable t) (evil-matchit) (form-feed) (highlight-indent-guides) (hungry-delete) (company-quickhelp) (company-jedi))
	unpin      (n/a)
	elpa       (n/a)
</pre></details>

**Comment by hlissner** [comment]
I'm inclined to fault `evil-buffer-new` for this. It steps around our buffer-switch hooks by using `set-window-buffer` to change the window's buffer instead of, say `switch-to-buffer`, which I believe is better suited to the task. I'm not sure what to do about this, but for now you can get around it with:

```lisp
(defadvice! +evil--new-buffer-a (orig-fn count file)
  :around #'evil-buffer-new
  (if file
      (funcall orig-fn count file)
    (with-current-buffer (switch-to-buffer (generate-new-buffer "*new*"))
      (funcall (default-value 'major-mode)))))
```

**Comment by hlissner** [comment]
This should be resolved as of 0bb4d4dfcb6de889b2c9201ab02d30eeafa5ea1c.
